[DO NOT MERGE] WIP: convert fxa-dev to use docker images

jrgm opened the pull request:
This branch converts fxa-dev to use docker images. Install time as low as 15 minutes.

Notes:
  - stops using cloudfront to serve static content
  - uses a localhost mysql image instead of RDS
  - drops roles kv,sync,able,log(kibana/ES), although may want to restore sync with a dockerized version
  - uses docker images for authdb, content, customs, oauth, and 123done (trusted and untrusted).
  - handles restart when a new docker image is pulled (or configuration changes)
  - tests still pass

TODO:
  - use profile docker now that https://github.com/mozilla/fxa-profile-server/pull/240 is merged
  - Dockerize and use for oauth-console, basket, auth-server & auth-mailer, sync
  - lots of cleanup of obsolete settings and obsolete yum installs still to come
  - drop the datadog integration if we're not using it?

MAYBE:
  - could pre-bake an AMI with initial yum dependencies, docker set up and pulled
  - could "invert" ansible run to have more run "on-box" (less network delay)
  - consolidate initial docker pulls

vladikoff commented on aws/docker-pull.sh:
```diff
@@ -0,0 +1,34 @@
+
+IMAGES=(
+  jrgm/123done
+  mozilla/fxa-auth-db-mysql
```
do we need a way to somehow specify the tags, (i.e picking one of these: https://hub.docker.com/r/mozilla/fxa-content-server/tags/ ), should we do this later or part of this PR?

jrgm replied:
Setting a tag is a TODO.

I was about to say that it should be fine to pull these here, but the largest layers in our images are the ones that will change most often. So this script also should be taught to handle tags. Good point.

jrgm replied:
Just to clarify, this script is only used to bootstrap. Updating of docker images (and container restart if changed) is handled by ansible (run in a periodic cronjob).

jrgm replied:
Actually, I'm calling this as good enough. It pulls the default `latest` image, then subsequently, when needed, it will pull and start the container with the configured tag. So, may be slightly slower to create a new `fxa-dev` box, but it will start with what is desired.

jrgm replied:
DONE

vladikoff commented on aws/environments/fxadevtest.yml:
```diff
@@ -0,0 +1,18 @@
+
+# bleh
+cdn_ssl_certificate_id: ASCAIAYDJPL5OT7LFEZSE
+content_static_resource_url: https://static-latest.dev.lcip.org
```
@jrgm are these using latest for static resources?

jrgm replied:
Usually values have default, but these ones don't so they have to set to something. Hence, bleh.

TODO: add defaults and drop these lines where we're not going to use Cloudfront.

jrgm replied:
DONE

vladikoff commented on roles/mysql/tasks/main.yml:
```diff
@@ -1,9 +1,33 @@
+    env:
+      MYSQL_ALLOW_EMPTY_PASSWORD: 1
+    volumes:
+      - /var/data/mysql:/var/lib/mysql
```
👍

vladikoff commented:
## Some good news:

```
AWS step: from 0 to  4m 35sec
update installed packages from 4m 35 to 5m 12s
auth 7m 23 -> 7 48 pull
```

## Bad news: 

I got this error because we removed `trustedJKUs`, need to update fxa-dev. This failed my deploy for now

> TASK [auth : generate public key pair] *****************************************
fatal: [ec2-54-187-25-121.us-west-2.compute.amazonaws.com]: FAILED! => {"changed": true, "cmd": ["docker", "run", "--rm", "--net=host", "-v", "/data/config/auth:/config", "-e", "CONFIG_FILES=/config/stage.json", "mozilla/fxa-auth-server", "node", "scripts/gen_keys.js"], "delta": "0:00:01.362750", "end": "2017-03-21 17:29:26.941074", "failed": true, "rc": 1, "start": "2017-03-21 17:29:25.578324", "stderr": "/app/node_modules/convict/lib/convict.js:524\n        throw new Error(errBuf);\n        ^\n\nError: configuration param 'trustedJKUs' not declared in the schema\n    at Object.rv.validate (/app/node_modules/convict/lib/convict.js:524:15)\n    at Object.<anonymous> (/app/config/index.js:681:6)\n    at Module._compile (module.js:409:26)\n    at Object.Module._extensions..js (module.js:416:10)\n    at Module.load (module.js:343:32)\n    at Function.Module._load (module.js:300:12)\n    at Module.require (module.js:353:17)\n    at require (internal/module.js:12:17)\n    at Object.<anonymous> (/app/scripts/gen_keys.js:26:16)\n    at Module._compile (module.js:409:26)", "stdout": "", "stdout_lines": [], "warnings": []}

vladikoff commented:
@jrgm pushed the JKU fixed, but getting SQS stack creation errors. Sent you logs

jrgm commented:
I cut a `docker` branch from `docker-changes` and pushed [that branch](https://github.com/mozilla/fxa-dev/compare/docker?expand=1) up. Ever merging back to master is not practical without breaking existing fxa-dev boxes. (This branch, `docker-changes`, should not be used anymore and closing this as a PR).


I rebased a bit to clean up some junk. I also added some ansible to poll the endpoints for `/__version__`, check that that worked, and dump the version info at the end.

So, with the completion time defined as when {auth,content,oauth,profile} are all up and running, five runs timed at 7:18, 7:22, 7:30, 7:36, 8:02. So 7 to 8 minutes, if no aws or other unexpected delays. (Note: oauth and profile are the last two roles to be started; just pulling fxa-dev and identity-pubkeys happen after that).

There's still some cleanup to do (dead ansible files not needed in the branch), and as @vladikoff pointed out, allowing docker tags to be used, and defaulting away the CDN config values. Other TODO items include:

- converting some of the services to set the config via environment variables (initial hacks just mounted the existing config files into the container).

- as code changes happen, older docker images will accumulate. I have a script (from spotify) that nicely garbage collects unused images, which can be run as a cronjob.

Longer term TODOs include:

- I used `host` networking since it easier, but it means that anything can talk to anything. It would be more correct to use `bridged` networking and only allow containers to talk to the services that they need to talk to.

- It would be interesting to experiment with removing linux capabilities from the containers to find the minimal set that they need. This would be to inform how production might be configured for security purposes.

jrgm commented:
See https://github.com/mozilla/fxa-dev/pull/301 for a `docker` branch of this work.
